feat: debugger related helpers and utils

aorumbayev opened the pull request:
## Proposed Changes

- Mostly implementing methods very similar to https://github.com/algorandfoundation/algokit-utils-py/pull/61
- Adding some refinements in docs generation


## TODO: 

- Smoke tests -> DONE
- Figure out the right places in app client to invoke simulate and persist utility -> DONE

tristanmenzel commented on src/types/debug-utils.ts:
```diff
@@ -0,0 +1,174 @@
+export const TEAL_SOURCEMAP_EXT = '.teal.tok.map'
+
+export interface ErrnoException extends Error {
+  errno?: number | undefined
```
The `| undefined` here is redundant if the field is marked as optional via the `?`

tristanmenzel commented on src/types/debug-utils.ts:
```diff
@@ -0,0 +1,174 @@
+  location: string
+  programHash: string
+
+  constructor(location: string, programHash: string) {
```
you can use the `private` or `public` keyword to prefix params in the constructor to create 'auto fields'.

eg. 
```
constructor(private location: string, private programHash: string) {
  // No need to copy the params now
}
```

tristanmenzel commented on src/debug-utils.spec.ts:
```diff
@@ -0,0 +1,151 @@
+
+      const result = AVMDebuggerSourceMap.fromDict(JSON.parse(await fs.readFile(sourcemapFilePath, 'utf8')))
+      for (const item of result.txnGroupSources) {
+        item.location = 'dummy'
```
Might be worth a comment explaining _why_ we're doing this - I assume it's just to stabilise the output across different machines.

tristanmenzel commented on src/types/debug-utils.ts:
```diff
@@ -0,0 +1,174 @@
+  }
+}
+
+export async function upsertDebugSourcemaps(sourcemaps: AVMDebuggerSourceMapEntry[], projectRoot: string): Promise<void> {
```
There's a bit of inconsistent casing here around sourcemaps and SourceMap

neilcampbell commented on src/types/debug-utils.ts:
```diff
@@ -0,0 +1,171 @@
+  syscall?: string
+}
+
+export interface AVMDebuggerSourceMapDict {
```
```suggestion interface AVMDebuggerSourceMapDict {
```

neilcampbell commented on src/types/debug-utils.ts:
```diff
@@ -0,0 +1,171 @@
+export const TEAL_FILE_EXT = '.teal'
+export const TEAL_SOURCEMAP_EXT = '.teal.tok.map'
+
+export interface ErrnoException extends Error {
```
```suggestion
interface ErrnoException extends Error {
```

neilcampbell commented on src/types/debug-utils.ts:
```diff
@@ -0,0 +1,171 @@
+  await fs.promises.writeFile(sourcesPath, JSON.stringify(sources.toDict()), 'utf8')
+}
+
+export async function writeToFile(filePath: string, content: string): Promise<void> {
```
```suggestion
async function writeToFile(filePath: string, content: string): Promise<void> {
```

neilcampbell commented on src/types/debug-utils.ts:
```diff
@@ -0,0 +1,171 @@
+  await fs.promises.writeFile(filePath, content, 'utf8')
+}
+
+export async function buildAVMSourcemap({
```
I don't think these exported functions really belong in the type directory.

aorumbayev replied:
@neilcampbell whats a better spot to place them in this case? I assume at the root of `src`?

neilcampbell commented on src/types/debug-utils.ts:
```diff
@@ -0,0 +1,171 @@
+  outputPath,
+  client,
+  withSources = true,
+}: {
```
Extract a type

aorumbayev replied:
extracted into separate type

aorumbayev commented on docs/capabilities/debugger.md:
```diff
@@ -0,0 +1,31 @@
+
+- `debug`: Indicates whether debug mode is enabled.
+- `projectRoot`: The path to the project root directory. Can be ignored if you are using `algokit-utils` inside an `algokit` compliant project (containing `.algokit.toml` file). For non algokit compliant projects, simply provide the path to the folder where you want to store sourcemaps and traces to be used with [`AVM Debugger`](links to extension). Alternatively you can also set the value via th ...
+- `traceAll`: Indicates whether to trace all operations. Defaults to false, this means that when debug mode is enabled, any (or all) application client calls performed via `algokit-utils` will store responses from `simulate` endpoint. These files are called traces, and can be used with [AVM Debugger](link to vscode extension) to debug TEAL source codes, transactions in the atomic group and etc.
```
Will defer this to detailed doc that will be on the extension repo. Changed the content here to remove and etc part with more description.

aorumbayev commented on src/debug-utils.ts:
```diff
@@ -0,0 +1,156 @@
+  projectRoot,
+  algod,
+  bufferSizeMb,
+}: {
```
done

aorumbayev commented on src/debug-utils.ts:
```diff
@@ -0,0 +1,156 @@
+  client,
+  withSources,
+}: {
+  sources: PersistSourceMapInput[]
```
done

aorumbayev commented:
@neilcampbell feel free to check again as time allows, im moving to templates/docs items for now

neilcampbell commented on src/debug-utils.ts:
```diff
@@ -0,0 +1,247 @@
+
+// === Internal methods ===
+
+export async function loadOrCreateSources(sourcesPath: string): Promise<AVMDebuggerSourceMap> {
```
```suggestion
async function loadOrCreateSources(sourcesPath: string): Promise<AVMDebuggerSourceMap> {
```

neilcampbell reviewed (approved):
You shouldn't need to export any of these.

robdmoore commented on docs/capabilities/debugger.md:
````diff
@@ -0,0 +1,31 @@
+```ts
+import { UpdatableConfig } from '@algorandfoundation/algokit-utils'
+
+const config = new UpdatableConfig()
````
I’m not sure this is right?

I would have thought it would be algokit.Config.configure(…)

robdmoore commented on docs/capabilities/debugger.md:
```diff
@@ -0,0 +1,31 @@
+
+- `debug`: Indicates whether debug mode is enabled.
+- `projectRoot`: The path to the project root directory. Can be ignored if you are using `algokit-utils` inside an `algokit` compliant project (containing `.algokit.toml` file). For non algokit compliant projects, simply provide the path to the folder where you want to store sourcemaps and traces to be used with [`AVM Debugger`](links to extension). Alternatively you can also set the value via th ...
+- `traceAll`: Indicates whether to trace all operations. Defaults to false, this means that when debug mode is enabled, any (or all) application client calls performed via `algokit-utils` will store responses from `simulate` endpoint. These files are called traces, and can be used with [AVM Debugger](link to vscode extension) to debug TEAL source codes, and a large variety of transaction types in ...
```
We should describe what the behaviour is when it’s false - it’s not clear

robdmoore commented on docs/capabilities/debugger.md:
```diff
@@ -0,0 +1,31 @@
+import { UpdatableConfig } from '@algorandfoundation/algokit-utils'
+
+const config = new UpdatableConfig()
+config.configure({ debug: true })
```
Probably a good idea to show the basic example and a full example?

robdmoore commented on docs/capabilities/debugger.md:
```diff
@@ -0,0 +1,31 @@
+
+Debugging utilities can be used to simplify gathering artifacts to be used with [AVM Debugger](link to vscode extension) in non algokit compliant projects. The following methods are provided:
+
+- `persistSourceMaps`: This method persists the sourcemaps for the given sources as AVM Debugger compliant artifacts.
```
I wonder if we should prefix these with a name prefix or even a different namespace like debugging?

aorumbayev replied:
i assume by namespace you aren't referring to typescript namespaces but es2015 module syntax ? 

In that case i've renamed the filename to `debugging.ts` and related docs

robdmoore commented on src/types/app-client.ts:
```diff
@@ -339,6 +342,18 @@ export class ApplicationClient {
     this._clearSourceMap = clearCompiled?.sourceMap
 
+    if (Config.debug && Config.projectRoot) {
+      persistSourceMaps({
```
This means we cal compile twice, is that a problem?

robdmoore replied:
Is also is probably missing the deploy time control parameters? Can we instead pass through the source map from this method rather than calling compile?

aorumbayev commented:
@robdmoore addressed your comments also propagated the fix for avoiding compilation twice to py version in https://github.com/algorandfoundation/algokit-utils-py/pull/61
